model/utils.go

```diff
@@ -72,10 +73,10 @@ func (sa StringArray) Equals(input StringArray) bool {
 	return true
 }
 
-var translateFunc goi18n.TranslateFunc = nil
+var translateFunc atomic.Value
```

Comment:
Do we really need this to be atomic? This looks like something that should be set before any `AppError` can be created and never touched again.
Wouldn't it be enough to move the call to `AppErrorInit()` further up in the server initialization function (before initializing things like jobs)?

Reply:
I agree. What made me think this needs to be atomic is that for this race itself, the setting of the value happens before the read code is executed. So I thought maybe there's something else going on. I couldn't actually repro the race locally. But I'll move it upwards to make it more robust.

Reply:
Yeah, looking at the race logs, it seems like `runServer` gets called at the same time from the two tests (`TestRunServerSuccess` and `TestRunServerSystemdNotification`). I don't think those would run in parallel, so maybe we are not properly waiting for every goroutine to be done on server shutdown? In any case, this race seems test specific as in production I'd assume we'd call that only once.

Reply:
Ok, I think I know what's happening. I moved the call to the top, but I still see this race happening in some other tests. I made a slight mistake by putting it too high. But I think I get the general problem here.

Since this is a global variable, so the value persists between multiple tests. So as a result, even if there is a slight mistiming where one test does not finish completely, and another test starts, this race will manifest. It's not just `TestRunServerSuccess` and `TestRunServerSystemdNotification`. If you see https://app.circleci.com/pipelines/github/mattermost/mattermost-server/18451/workflows/f3065425-eba6-412e-8caf-10c897633637/jobs/105295 and https://app.circleci.com/pipelines/github/mattermost/mattermost-server/18451/workflows/f3065425-eba6-412e-8caf-10c897633637/jobs/105294, there are multiple failures of the same race.

All of it is because of consecutive tests running in the same process. The actual code is fine. I am thinking of covering the set call with a `sync.Once`, so that it's only set once during the lifetime of a process. That does not go the full way of guarding a variable, but should still fix the race. Thoughts?

The issue is more complicated by this `utils.TranslationsPreInit` which sets yet another global variable, and then we set that global variable to another variable in model package. Talk about complexity :)
```
if err := utils.TranslationsPreInit(); err != nil {
		return nil, errors.Wrapf(err, "unable to load Mattermost translation files")
	}
	model.AppErrorInit(utils.T)
```

Reply:
hmm interesting findings. I would agree that `sync.Once` will fix the race, we must ensure that we are not copying that global var anywhere though.

Reply:
> Since this is a global variable, so the value persists between multiple tests. So as a result, even if there is a slight mistiming where one test does not finish completely, and another test starts, this race will manifest. It's not just `TestRunServerSuccess` and `TestRunServerSystemdNotification`. If you see https://app.circleci.com/pipelines/github/mattermost/mattermost-server/18451/workflows/f3065425-eba6-412e-8caf-10c897633637/jobs/105295 and https://app.circleci.com/pipelines/github/mattermost/mattermost-server/18451/workflows/f3065425-eba6-412e-8caf-10c897633637/jobs/105294, there are multiple failures of the same race.

yes, somehow we are not guaranteeing order between accesses.

> All of it is because of consecutive tests running in the same process. The actual code is fine. I am thinking of covering the set call with a `sync.Once`, so that it's only set once during the lifetime of a process. That does not go the full way of guarding a variable, but should still fix the race. Thoughts?

Sounds good to me :+1:

Reply:
Ok, pushed that. Will let it run through `-race` flag to confirm it fixes the race.

Reply:
Worked.